### DISCUSSION TITLE
How to trigger a full refresh of dbt incremental models with software-defined assets

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#8873] [Q&A] [answered]
Hello, when using [|DBT software-defined-asset](https://docs.dagster.io/_apidocs/libraries/dagster-dbt#assets), is there a way to trigger a [full refresh](https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models#how-do-i-rebuild-an-incremental-model) of my incremental models ?

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1657669495255529?thread_ts=1657669495.255529&cid=C01U954MEER).*

**Comment by yuhan** [comment]
When configuring your `dbt_cli_resource`, you can pass in `{... (other config options) ..., "full-refresh": True}`. For example,

```python
dbt_cli_resource.configured({... (other config options) ..., "full-refresh": True})
```

**Reply by sryza** [reply]
@OwenKephart - thoughts on whether there's a way to make this more easily discoverable?

**Comment by juniorb2ss** [comment]
If you need to pass the argument to `dbt_cli.run` command you can do this:

```
context.resources.dbt_cli.run(...,full_refresh=True)
```

> ... kwargs are passed in as additional parameters.
Reference: https://docs.dagster.io/_apidocs/libraries/dagster-dbt#dagster_dbt.DbtCliResource.run

**Comment by remigabillet** [comment]
I'm also using dbt software-defined assets.  I load assets with `load_assets_from_dbt_project()`, so I don't interact with the `dbt_cli` resource at all. 

I would like to trigger a full_refresh directly from the Dagit UI. Any ideas?

**Reply by sryza** [reply]
@rexledesma - do you have tips on how to do this?

**Reply by rexledesma** [reply]
https://github.com/dagster-io/dagster/discussions/8873#discussioncomment-4975321

**Comment by rexledesma** [comment, marked as answer]
When materializing your assets, you can pass configuration to the underlying op that backs your software-defined assets. it's named something like `run_op_XXX`.

For example, the op that backs my dbt software-defined assets is named `run_dbt_b83ee_a92a1`. You can pass this configuration in the launchpad to ensure that a full refresh occurs.:

```
ops:
  run_dbt_b83ee_a92a1:
    config:
      full_refresh: true
```

**Reply by remigabillet** [reply]
Thanks for your help Rex. I did not know you had the `full_refresh` config argument implemented. It's great!

I still need help finding where to enter these options on the Dagit UI. Here's what I am doing.

- I select assets I want to full-refresh from the Assets DAG
- I shift+click "Materialize selected"
- the overlay titled "Launchpad (configure assets)" is displayed
- I'm able to pass config to each asset but I'm not seeing what the upcoming dbt run ops are called

I would also like to `define_asset_job` with `full_refresh` enabled. Is that possible?

**Reply by rexledesma** [reply]
> Thanks for your help Rex. I did not know you had the `full_refresh` config argument implemented. It's great!
> 
> I still need help finding where to enter these options on the Dagit UI. Here's what I am doing.
> 
> * I select assets I want to full-refresh from the Assets DAG
> * I shift+click "Materialize selected"
> * the overlay titled "Launchpad (configure assets)" is displayed
> * I'm able to pass config to each asset but I'm not seeing what the upcoming dbt run ops are called

You should ensure that at least one of your dbt assets are selected. Typing in `op:` in the launchpad should display the typeahead to show the underlying dbt op.

> I would also like to `define_asset_job` with `full_refresh` enabled. Is that possible?

`define_asset_job` takes in a `config` parameter that you can use to bake this configuration in with the asset job.

**Reply by remigabillet** [reply]
Thanks Rex, I figured it out by typing `run_dbt_` in the launchpad to generate all dbt-related ops. This also displays which dbt ops are irrelevant to the job.

<img width="574" alt="Screenshot 2023-02-16 at 8 11 16 AM" src="https://user-images.githubusercontent.com/20303/219318138-055fc239-25a5-45ec-8835-caab8de398ab.png">

With this approach, I can pass the `full_refresh` config to each of the ops when using dagit.

In the context of an asset job, I can pass full_refresh to all ops this way:
![Screenshot 2023-02-16 at 10 11 46 AM](https://user-images.githubusercontent.com/20303/219320189-8d99c79d-94bc-4096-a879-f2af94073ac7.png)


I have 2 follow-up questions:
- I'm working in a dev environment, but I would like to know if the same op names will be generated on production. My dbt asset keys are different btw dev & prod (they use the output schema). How deterministic are these dbt op names?
- While this works, this approach feels brittle. if I adjust my assets and their dependencies, the related dbt ops could change right? This means that for all asset jobs relying on dbt, I need to go through this Launchpad and manually put the jobs' configs in code. Have you explored a different way to pass an argument to all ops? I've been using run tags to broadcast an argument to all assets & ops, maybe there's a better way.

**Reply by remigabillet** [reply]
- I answered the first question myself by deploying to production. The op names are consistent between environments, so the `define_asset_job` call works well.
- I would still like to know if there is an alternate approach to pass full_refresh to all ops within a job.

**Reply by rexledesma** [reply]
> I'm working in a dev environment, but I would like to know if the same op names will be generated on production. My dbt asset keys are different btw dev & prod (they use the output schema). How deterministic are these dbt op names?

The dbt op name is determined by a function of the dbt project id, as well as some of its configuration. It is deterministic, but should probably be accessed programmatically, rather than encoding its raw string directly. Something like:

```
dbt_assets = load_assets_from_dbt_manifest(...)
op_name = dbt_assets[0].op.name
```

**Reply by remigabillet** [reply]
thanks for the follow-up Rex. This is much better than hardcoding the op name.
